runtime/src/accounts_db.rs

```diff
@@ -3065,56 +3107,148 @@ impl AccountsDb {
+                let scan_result: ScanStorageResult<Pubkey, DashSet<Pubkey>> = self
+                    .scan_account_storage(
+                        *remove_slot,
+                        |loaded_account: LoadedAccount| Some(*loaded_account.pubkey()),
+                        |accum: &DashSet<Pubkey>, loaded_account: LoadedAccount| {
+                            accum.insert(*loaded_account.pubkey());
+                        },
+                    );
+                scan_storages_elasped.stop();
+                purge_stats
+                    .scan_storages_elasped
+                    .fetch_add(scan_storages_elasped.as_us(), Ordering::Relaxed);
+
+                let mut purge_accounts_index_elapsed =
+                    Measure::start("purge_accounts_index_elapsed");
+                // Purge this slot from the accounts index
+                let purge_slot: HashSet<Slot> = vec![*remove_slot].into_iter().collect();
+                let mut reclaims = vec![];
+                match scan_result {
+                    ScanStorageResult::Cached(_) => {
+                        panic!("Should not see cached keys in this `else` branch, since we checked this slot did not exist in the cache above");
+                    }
+                    ScanStorageResult::Stored(stored_keys) => {
+                        for set_ref in stored_keys.iter() {
+                            self.accounts_index.purge_exact(
+                                set_ref.key(),
+                                &purge_slot,
+                                &mut reclaims,
+                            );
+                        }
+                    }
+                }
+                purge_accounts_index_elapsed.stop();
+                purge_stats
+                    .purge_accounts_index_elapsed
+                    .fetch_add(purge_accounts_index_elapsed.as_us(), Ordering::Relaxed);
+
+                // `handle_reclaims()` should remove all the account index entries and
+                // storage entries
+                let mut handle_reclaims_elapsed = Measure::start("handle_reclaims_elapsed");
+                // Slot should be dead after removing all its account entries
+                let expected_dead_slot = Some(*remove_slot);
+                self.handle_reclaims(
+                    &reclaims,
+                    expected_dead_slot,
+                    Some(purge_stats),
+                    Some(&mut ReclaimResult::default()),
+                    false,
+                );
+                handle_reclaims_elapsed.stop();
+                purge_stats
+                    .handle_reclaims_elapsed
+                    .fetch_add(handle_reclaims_elapsed.as_us(), Ordering::Relaxed);
+                // After handling the reclaimed entries, this slot's
+                // storage entries should be purged from self.storage
+                assert!(self.storage.get_slot_stores(*remove_slot).is_none());
```

Comment:
ick, this can actually race with `clean_accounts()`. If clean removes some of the the accounts index entries first, then we wont remove them here, and the `handle_reclaims()` call here may not remove all the storage entries (clean may have the remaining reclaims from the index necessary to mark the slot as dead)

Reply:
> Luckily right now AccountsBackgroundService runs all Bank::drop() serially with clean_accounts()

With the upcoming duplicate slot thing, how about introducing a runtime check to really validate this?

I think we can signal AccountsDb here at the identical timing to setup callbacks for all banks: https://github.com/solana-labs/solana/blob/44831c18d29f5c43feb0ffdb0abe258be34d4a99/core/src/tvu.rs#L237-L240

so that all subsequent `purge_slot()` must be coming from the ABS thread, not directly from `Bank::drop()`-ed thread.

Reply:
i mean, I think this property could tend to be broken in the future by other code juggling. :)

Reply:
@ryoqun, hmm so currently `Bank::new_from_parent()`, already guarantees that all child banks will have the proper callback.

>I think we can signal AccountsDb here at the identical timing to setup callbacks

So something to verify from `purge_slot()` that the caller is ABS thread, i.e. something like checking the thread id is equal to the ABS thread?

Reply:
> So something to verify from purge_slot() that the caller is ABS thread, i.e. something like checking the thread id is equal to the ABS thread?

@carllin Oh, I was imaging more simpler way than thread id like this:

```rust
let callback = accounts_db.create_drop_callback(); // internally set `.drop_callback_installed` to `true`.
// Before replay starts, set the callbacks in each of the banks in BankForks 
for bank in bank_forks.read().unwrap().banks().values() { 
    bank.set_callback(callback)
}


impl AccountsDb {
   fn purge_slot(from_abs: bool) {
     if self.drop_callback_installed && !from_abs {
        panic!("bad drop callpath detected; bank.drop must be serialized")
     }
   }
}

impl Drop for Bank {
  fn drop() {
     accounts_db.purge_slot(..., false)
  }
}

impl ABS {
    accounts_db.purge_slot(..., true);
}
```

Reply:
anyway this is optional nicety.